### PR TITLE
add 2k chunk

### PR DESCRIPTION
**Pull request by jihyun-jeong-854**

### CODE REVIEW

File: system/flcore/servers/serverbase.py [hunk 1]
```diff
@@ -5,11 +5,13 @@
 import copy
 import time
 import random
+from sklearn.preprocessing import label_binarize
+from sklearn import metrics
```
Comment (albert-no) [inline system/flcore/servers/serverbase.py:9]: 이건 사용하는건가요?

File: system/flcore/servers/serverbase.py [hunk 2]
```diff
@@ -62,28 +64,31 @@ def __init__(self, args, times):
         self.new_clients = []
         self.eval_new_clients = False
         self.fine_tuning_epoch = args.fine_tuning_epoch
-        
+
+
```
Comment (albert-no) [inline system/flcore/servers/serverbase.py]: 빈칸은 굳이 안늘리겠습니다. black formatter를 사용해보기 바랍니다

File: system/flcore/servers/serverbase.py [hunk 3]
```diff
@@ -154,6 +159,70 @@ def aggregate_parameters(self):
...
+        min_ = self.get_min_grad(gradient)
+        # min_ = 1e-9
+        
+        for g in gradient:
+            g.data = torch.where(torch.abs(g.data) >= min_, g.data, 0.)
+            
+        return gradient 
+    
+    def get_min_grad(self,gradient):
+        all_grads = torch.empty(0,dtype=torch.float32).to(self.device)
```
Comment (albert-no) [inline system/flcore/servers/serverbase.py]: all_grads초기화가 필요한가요?

File: system/flcore/servers/serverbase.py [hunk 4]
```diff
@@ -154,6 +159,70 @@ def aggregate_parameters(self):
...
+        for grad in gradient:
+            all_grads = torch.cat([all_grads,grad.reshape(-1)])
```
Comment (albert-no) [inline system/flcore/servers/serverbase.py]: all_grads = torch.cat([grad for grad in gradient])
이런식이 나을거같아요
어차피 전체 min을 찾는거라 굳이 reshape안해도될거같구요

File: system/flcore/servers/serverbase.py [hunk 5]
```diff
@@ -154,6 +159,70 @@ def aggregate_parameters(self):
...
+        for grad in gradient:
+            end_idx = start_idx + grad.data.numel()
+            topk_chunk_grad.append(torch.Tensor(topk_chunk_grad_flattened[start_idx:end_idx]).view(grad.data.shape))
+            start_idx = end_idx
+            
+        return topk_chunk_grad    
+    
+    def global_topk(self,gradient):
+        min_ = self.get_min_grad(gradient)
+        # min_ = 1e-9
```
Comment (albert-no) [inline system/flcore/servers/serverbase.py]: 지난번에 테스트용으로 두었던건 지우겠습니다

File: system/flcore/servers/serverbase.py [hunk 6]
```diff
@@ -154,6 +159,70 @@ def aggregate_parameters(self):
...
+            
+        optimizer = torch.optim.SGD(self.global_model.parameters(), lr = self.args.local_learning_rate)    
+        optimizer.step()
+  
+    def chunk_topk(self,gradient):
+        all_grads = torch.empty(0,dtype=torch.float32).to(self.device)
+        for grad in gradient:
+            all_grads = torch.cat([all_grads,grad.reshape(-1)])
+            
+        chunks = torch.chunk(all_grads,2*self.args.topk,dim = -1)
```
Comment (albert-no) [inline system/flcore/servers/serverbase.py]: 2k를 쓰는건 예를 든거여서요
chunk topk알고리즘은 chunk의 개수를 입력으로 받아서 각 chunk별 1등을 뽑는알고리즘으로 독립적으로 정의하는게 좋겠습니다

File: system/flcore/servers/serverbase.py [hunk 7]
```diff
@@ -154,6 +159,70 @@ def aggregate_parameters(self):
...
+        optimizer.step()
+  
+    def chunk_topk(self,gradient):
+        all_grads = torch.empty(0,dtype=torch.float32).to(self.device)
+        for grad in gradient:
+            all_grads = torch.cat([all_grads,grad.reshape(-1)])
+            
+        chunks = torch.chunk(all_grads,2*self.args.topk,dim = -1)    
+        for chunk in chunks:
+            local_max = torch.abs(chunk.data).max()
```
Comment (albert-no) [inline system/flcore/servers/serverbase.py]: 아마 argmax같은게 있을거에요

**Review by albert-no — COMMENTED** [review]
어차피 gradient가 아니라 학습전후 parameter의 차이를 사용하는 코드로 수정할 계획이지만
코드의 뼈대는 그대로될것같아요.
이후 코드 작업할때 코멘트 확인하며 수정하기 바랍니다.
특히 global_topk와 chunk_topk를 각각 독립적으로 구분하는게 좋겠습니다.

File: system/flcore/servers/serverbase.py (hunk 1 above)
Comment (rachel618) [reply system/flcore/servers/serverbase.py:9]: 자세히 보니 label_binarize는 기존 코드에 num_class == 2 일 때만 사용하도록 되어있어서 지우겠습니다.
 sklearn 의 metrics는  auc score 계산할 때 사용합니다!

File: system/flcore/servers/serverbase.py (hunk 2 above)
Comment (jihyun-jeong-854) [reply system/flcore/servers/serverbase.py]: 네 알겠습니다!

File: system/flcore/servers/serverbase.py (hunk 4 above)
Comment (jihyun-jeong-854) [reply system/flcore/servers/serverbase.py]: 이렇게 바꾸면 초기화도 필요 없을 것 같습니다. reshape도 안하도록 수정하겠습니다!

File: system/flcore/servers/serverbase.py (hunk 5 above)
Comment (jihyun-jeong-854) [reply system/flcore/servers/serverbase.py]: 네 알겠습니다.
